example/contracts/Migrations.sol

```diff
@@ -0,0 +1,21 @@
+contract Migrations {
+  address public owner;
+  uint public last_completed_migration;
+
+  modifier restricted() {
+    if (msg.sender == owner) _
+  }
+
+  function Migrations() {
+    owner = msg.sender;
+  }
+
+  function setCompleted(uint completed) restricted {
+    last_completed_migration = completed;
+  }
+
+  function upgrade(address new_address) restricted {
+    Migrations upgraded = Migrations(new_address);
+    upgraded.setCompleted(last_completed_migration);
+  }
```

Comment:
Good points. The idea is that you can deploy a different Migrations contract as long as it contains the same signature as the old one for two specific functions, `setCompleted()` and `last_completed_migration()`, the latter provided by solidity. If this is true, the implementation of the new one does not matter: during an upgrade, we'll just treat it like its the old one and call its `setCompleted()` function, which must exist. The `data` field that gets generated for that transaction should be the same. 

You have very good points though. This is actually something I added initially, but have not tested. Perhaps I can write a test for it.